tracer/src/Datadog.Trace/ClrProfiler/Instrumentation.cs

```diff
@@ -101,6 +101,21 @@ public static void Initialize()
                 Log.Error(ex, "CallTarget state ByRef cannot be enabled: ");
             }
 
+            try
+            {
+                Log.Debug("Initializing TraceAttribute instrumentation.");
+                var payload = InstrumentationDefinitions.GetTraceAttributeDefinitions();
+                NativeMethods.AddTraceAttributeInstrumentation(payload.DefinitionsId, payload.AssemblyName, payload.TypeName);
+                Log.Information("TraceAttribute instrumentation enabled with Assembly={AssemblyName} and Type={TypeName}.", payload.AssemblyName, payload.TypeName);
+            }
+            catch (Exception ex)
+            {
+                Log.Error(ex, ex.Message);
+            }
+
```

Comment:
Why did this need to move? 🤔

Reply:
Service Discovery uses the HTTP client and I need to use it before creating the tracer instance to decide if we use the CIVisibilityProtocol writer or the APM one, also if the Intelligent Test Runner can be enabled or not. For that reason I moved the tracer creation (Managed Side Initialization) before the PInvoke call with the integrations list (so the ServiceDiscovery HTTP call doesn't get instrumented).

What I saw is that if I do that the `TraceAttribute` instrumentation breaks (don't know why yet) so I have to keep the call for the TraceAttribute before the tracer creation.

Reply:
I have plan to deep dive into this issue with the PR to reduce the number of PInvoke calls in the future.

Reply:
Oh man, this is the worst 😩 We've had bugs around this sort of thing before (it was telemetry last time) where we initialize things to early essentially.

Feels like we need a more systematic approach somehow, to avoid initialization bugs creeping in, but I don't really know _how_ we could do that😟

Reply:
Yeah I know, For me,  we should check if the tracer is properly initialized before any usage (maybe we can do that in the `CallTargetInvoker`). That way it will be safe to do the PInvokes at the begining.

Reply:
> What I saw is that if I do that the TraceAttribute instrumentation breaks (don't know why yet) so I have to keep the call for the TraceAttribute before the tracer creation.

Nothing comes to mind, but I can dig into this issue. We should ship what you have and not block on this issue though

Reply:
Yes, we can do that in another PR. What I found is, when I initialize the Tracer before TraceAttribute instrumentation, the TraceAttribute Samples app doesn't create the root span (the one with the TraceAttribute). Maybe is something to do with the async Main in the app. 🤷🏻

Reply:
I kind of prefer to execute all the PInvokes calls first but instrumentation disabled until we initialize the tracer. That way we ensure all rewrites are done earlier but without interfering in the tracer initialization.